src/parse/parse_step.go

```diff
@@ -64,6 +64,31 @@ func parse(tid int, state *core.BuildState, label, dependent core.BuildLabel, fo
 		if err := subrepo.LoadSubrepoConfig(); err != nil {
 			return err
 		}
+
+		// Validate plugin ID is the same as the subrepo name
+		if pluginID := state.RepoConfig.PluginDefinition.Name; pluginID != "" {
+			subrepoName := subrepo.Name
+			if subrepo.Arch.String() != "" {
+				subrepoName = strings.TrimSuffix(subrepo.Name, "_"+subrepo.Arch.String())
+			}
+			if !strings.EqualFold(pluginID, subrepoName) {
+				return fmt.Errorf("Subrepo name %q should be the same as the plugin ID %q", subrepoName, pluginID)
+			}
+		}
+
+		// Validate the plugin config keys set in the host repo
+		definedKeys := map[string]bool{}
+		for key, definition := range state.RepoConfig.PluginConfig {
+			configKey := getConfigKey(key, definition.ConfigKey)
+			definedKeys[configKey] = true
+		}
+		if plugin := state.Config.Plugin[subrepo.Name]; plugin != nil {
+			for key := range plugin.ExtraValues {
+				if _, ok := definedKeys[strings.ToLower(key)]; !ok {
+					return fmt.Errorf("Unrecognised config key %q for plugin %q", key, subrepo.Name)
+				}
+			}
+		}
```

Comment:
Can we move this into it's own function? This top level function is quite big already.